byway/src/main/webapp/home.js

```diff
@@ -110,9 +152,16 @@ function calcRoute(directionsService, directionsRenderer, start, end, waypoints)
   directionsService.route(request, function(response, status) {
     if (status == 'OK') {
       directionsRenderer.setDirections(response);
-    } else {
+    } 
+    else if (status === google.maps.DirectionsStatus.OVER_QUERY_LIMIT) {
+      delayPromise(1000)
+      .then(() => calcRoute(directionsService,directionsRenderer,start,end,waypoints));
```

Comment:
Writing this as a recursive callback loop makes the logic harder to read and understand, and potentially harder to control, e.g. it's difficult to set a limit to the number of attempts that will be made. It's also impossible for the caller to tell if the function has completed or not, since calling `calcRoute` doesn't wait for the route to be calculated in any way.

I would recommend making a function that returns a promise of the result of `directionService.route`, and write `calcRoute` as an async function with a loop that retries the call to the directions service until it works (or some predetermined attempt limit is reached).

Reply:
done